### ISSUE TITLE
ShapeCast3D detects colliders behind it 

### ISSUE THREAD

**Issue by Sheepy3** [godotengine/godot#93880]
### Tested versions

- Reproducible in 4.2.2 Stable, 4.1.4 Stable, 4.0.0 Stable, 4.3 Beta 2. 

### System information

Godot v4.2.2.stable - Windows 10.0.19045 - Vulkan (Forward+) - dedicated NVIDIA GeForce RTX 3090 (NVIDIA; 31.0.15.3623) - AMD Ryzen 9 5900X 12-Core Processor (24 Threads)

### Issue description

I have a sweeping shapecast3D attached to an "agent node". this shapecast3D is intended to detect collisions within the shapecast3D. However, the shapecast3D is detecting objects directly behind it. 

https://github.com/godotengine/godot/assets/109123711/1f9d4f1b-1b49-481a-a245-2b68d2db1f4b

### Steps to reproduce

1) open the project 
2) run the main scene
3) observe shapecast3D collision behavior

### Minimal reproduction project (MRP)

[shapecastminimalrepro.zip](https://github.com/user-attachments/files/16076444/shapecastminimalrepro.zip)

**Comment by kleonc** [comment]
I don't think that's a bug, the casted BoxShape has `size.x = 100` and the box is centered, hence from the starting position it reaches `-50 <= x <= 50`. The "agent behind" is roughly `2.69722` units away on `x`, hence it collides with the casted box. Setting `size.x = 5.3` (less than `2 * 2.69722`) for the casted BoxShape makes it no longer collide, as expected.

**Comment by Calinou** [comment]
> What might be confusing is how the debug collision is drawn:

Is there a way we could improve ShapeCast3D debug drawing to be less confusing?

**Comment by Sheepy3** [comment]
>Is there a way we could improve ShapeCast3D debug drawing to be less confusing?
from my understanding, it seems like the debug drawing doesn't actually represent the true shape of the shapecast3D? I would presume the debug collision would be drawn in a manner which shows what region is being checked for collisions, no?

**Comment by Sheepy3** [comment]
heres a video comparing the agent's shapecast with the default target position and with a changed target position (ive also attached this project specifically if thats helpful at all)

https://github.com/godotengine/godot/assets/109123711/48121bd2-87a0-4983-b342-7e78974fea83


[shapecastminimalrepro_2.zip](https://github.com/user-attachments/files/16091819/shapecastminimalrepro_2.zip)

**Comment by Griiimon** [comment]
> Shouldn't these return `0.0` in such case? 🤔 I mean there's no "safe" movement as we start colliding but returning `0.0` indicating we can't move at all makes more sense to me than returning `1.0` indicating that we can move to `ShapeCast3D.target_position` without any collisions.

Agreed! this sounds very much like a bug and should probably get its own issue.
